cassandra/cqltypes.py

```diff
@@ -444,6 +444,8 @@ def deserialize(byts, protocol_version):
 
     @staticmethod
     def serialize(dec, protocol_version):
+        if (Decimal(dec).is_nan()):
+            raise Exception("NaN is not a valid Decimal value")
```

Comment:
So I think this is happening just a bit too early.

Note that immediately after this operation you can see serialize(...) trying to call as_tuple() on the input "dec" value.  If it's a Decimal type this'll work great, but if it's not you'll chuck the AttributeError.  In that case the code will try to convert "dec" to a Decimal (via Decimal(dec)) and try as_tuple() on the result.

I'd suggest that your code should follow a similar pattern.  Call dec.is_nan() and throw as necessary, then call is_tuple() and continue normally.  If any of that throws an AttributeError do something similar in the exception handler; convert "dec" to a Decimal and then do the is_nan() + is_tuple() ops.